### ISSUE TITLE
Bug: after re-create solid-component custom element, setState triggers unrelated computation node

### ISSUE THREAD

**Issue by Meigyoku-Thmn** [ryansolid/solid#17]
I don't know this bug is in solid-js or solid-component so I just report in here.
This is my sample project: [solid-js-bug.zip](https://github.com/ryansolid/solid/files/2999653/solid-js-bug.zip)
So in my attached folder, in main.jsx, there is the code like this:
```jsx
const componentMap = ['sub-element', 'div']
const App = (props, { element }) => {
   console.log('Create App')
   const [state, setState] = createState({
      selectedCompIndex: 0,
   });
   const ComponentSwicher = componentSwitcher();
   return (<div style="background-color: yellow">
      <p>App Component</p>
      <p>Selected Component Index: {(state.selectedCompIndex)}</p>
      <button onClick={() => setState({ selectedCompIndex: 0 })}>Select Component 0</button>
      <button onClick={() => setState({ selectedCompIndex: 1 })}>Select Component 1</button> <br />
      <ComponentSwicher id="component" _tag={() => componentMap[state.selectedCompIndex]} />
   </div>);
}
```
ComponentSwicher:
```jsx
componentSwitcher = () => ({
   _tag: tagGetter = () => null,
   _is: builtInGetter = () => null,
   onInit = () => { },
   ...restProps,
}) => {
   let element;
   if (typeof tagGetter != 'function') tagGetter = ((value) => () => value)(tagGetter);
   if (typeof builtInGetter != 'function') builtInGetter = ((value) => () => value)(builtInGetter);
   if (tagGetter() == null) return null;
   r.wrap(() => {
      if (element == null)
         element = document.createElement(tagGetter(), { is: builtInGetter() })
      else {
         let newElement = document.createElement(tagGetter(), { is: builtInGetter() });
         element.parentNode.replaceChild(newElement, element);
         element = newElement;
      }
      r.spread(element, () => restProps);
      onInit(element);
   });
   return element;
}
```
The idea is that I can switch an element to another element, in this exemple, ComponentSwicher can change between sub-element and div.
This is the sub-element component:
```jsx
const SubElement = (props, { element }) => {
   console.log('Create SubElement')
   const [state, setState] = createState({
      counter: 0,
   });
   return (<div style="background-color: rgb(200, 200, 200)">
      <p>SubElement Component</p>
      <button onClick={() => setState({ counter: state.counter + 1 })}>Increase counter</button>
      <widget-element counter={(state.counter)}
         onCounterChanged={e => setState({ counter: e.detail.counter })} />
   </div>);
}
```
And widget-element component:
```jsx
const WidgetElement = (props, { element }) => {
   console.log('Create WidgetElement')
   let counter = props.counter; // deleting this line will bypass the bug, but I need to keep this line in my project
   return (<div style="background-color: rgb(255, 204, 153)">
      <p>WidgetElement Component</p>
      <p>Counter prop: {(props.counter)}</p>
      <button onClick={() => {
         element.dispatchEvent(new CustomEvent("counterchanged", {
            bubbles: true, composed: true, detail: { counter: props.counter + 1 }
         }));
      }}>Increase counter</button> <br/><br/>
   </div>);
}
```
If you click "Select Component 1", then click "Select Component 0", then click any of "Increase Counter" buttons, instead of increasing the counter, it just re-create the entire sub-element component (you can see my log in console).
If you remove this line "let counter = props.counter;" in widget-element, the bug will disappear.

**Comment by ryansolid** [comment]
Cursory glance I don't see what the issue is. Where you have that props statement should be inert (non-tracking area) as all Web Components create their own root. It could be that I use r.insert in the component library to do that insert and if somehow what is getting returned from that component is a function. r.insert will auto wrap functions. I just don't see it from the code.

I've downloaded the project to take a look at it. I'm hitting some weird path issues with the ES module shim stuff (perhaps with me being under WSL in windows... it's always a bit weird). I'm going to take a few mins to set this up with a packager I'm more familiar with and see if I can debug the issue.

**Comment by ryansolid** [comment]
Ok I can definitely reproduce it. It is that it is tracking at that point. I had a misunderstanding of how S.js worked. Because of the scenarios I had worked on I assumed roots were inert. They are not. I just never needed to grab a value outside of an effect that I never noticed. I will push out some fixes. Most notably to Solid Components and Babel Plugin ('each' control is fine but the others are at risk of this issue).  Another scenario to add to the tests. Thank you. I will update you when I've pushed all the new versions.
